# 1获取数据

## 1.1数据集初步了解

该原始数据集记录了一家英国在线零售公司在2010.12.01~2011.12.09期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的
礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间,即交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

# 2读取数据

导入数据分析所需要的库，并通过panda的read_csv函数，将原始数据"e_commerce.csv"里的数据内容，解析为DataFrame，并赋值给变量`original_data`

In [74]:
import pandas as pd

In [75]:
original_data = pd.read_csv(r"C:\Users\11485\Desktop\e_commerce.csv")

In [76]:
original_data.head() #显示该原始数据集的表头

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

# 3明确**分析目标**

对于此数据集分析的目标是：根据市场销售数据、挖掘畅销产品，以便制定更有效的市场策略来提升营收。

# 4评估数据

在这一部分，我将对上一部分建立的`original_data`这个DataFrame所进行的数据进行评估。  
评估主要从两个方面进行：结构和内容，即整齐度和干净度。数据的结构性问题是指不符合“每一列是一个变量(属性/特征)，没一行是一个观察值(对象/示例)，每一个单元格
是一个值(可为字符串、数字、……)”。数据的内容性问题包括存在丢失数据、重复数据、无效数据等。

## 4.1评估数据整齐度

In [77]:
original_data.sample(10) #显示该数据集的前10行

InvoiceNo StockCode                    Description  Quantity  \
241747    558240     22371   AIRLINE BAG VINTAGE TOKYO 78         1   
353216    567793     22625             RED KITCHEN SCALES         2   
88724     543814     22972        CHILDREN'S SPACEBOY MUG         6   
440765    574547    85099B        JUMBO BAG RED RETROSPOT         6   
479106    577140     23581         JUMBO BAG PAISLEY PARK        10   
21270     538071    85035C  ROSE 3 WICK MORRIS BOX CANDLE         1   
134757    547867     82600          NO SINGING METAL SIGN         1   
59168     541274     21670  BLUE SPOT CERAMIC DRAWER KNOB        12   
430725    573585     22621     TRADITIONAL KNITTING NANCY        19   
141200    548504    15056p         EDWARDIAN PARASOL PINK         1   

             InvoiceDate  UnitPrice  CustomerID         Country  
241747   6/27/2011 15:27       4.13         NaN  United Kingdom  
353216   9/22/2011 11:40       8.50     12657.0          France  
88724    2/13/2011 15:17       1.65     15640.0  United Kingdom  
440765   11/4/2011 15:08       2.08     17841.0  United Kingdom  
479106   11/18/2011 9:10       2.08     13269.0  United Kingdom  
21270    12/9/2010 14:09       8.47         NaN  United Kingdom  
134757   3/27/2011 13:56       2.10     13357.0  United Kingdom  
59168    1/17/2011 11:24       1.25     15696.0  United Kingdom  
430725  10/31/2011 14:41       3.29         NaN  United Kingdom  
141200   3/31/2011 14:33      12.46         NaN  United Kingdom

从抽样的10行数据来看，数据符合“每列是一个变量，每行是一个观察值，每单元格是一个值”的原则，具体来看每行是关于某商品的一次交易，每列是交易相关的各个变量，因此不存在结构性问题。

## 4.2评估数据干净度

In [78]:
original_data.info() # 显示该数据集有多少变量(列)，每列变量有多少空缺值，以及数据类型是什么？

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


- 该数据集共有541909条数据(行)。

- Description存在541909-540455=1454个空缺值，CustmoerID存在541908-406829=135079个空缺值。

- InviceData数据类型为字符串，但是应该为时间。

- CustomerID数据类型为浮点数，但是应该为字符串。

### 4.2.1评估缺失数据

对原始数据集筛选出变量Description(列)为空值的观测值(行)

In [79]:
original_data[original_data["Description"].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

对于`Description`变量(列)，有1454个观测值(行)缺失变量值(其对应单元格为空)

从输出结果来看，这些缺失`Description`变量值的观测值，`UnitPrice`变量值可能也为零。为了验证猜想，我们增加筛选条件，看看是否存在`Description`变量缺失且`UnitPrice`不为零的观测值。

In [80]:
original_data[(original_data["Description"].isnull()) & (original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

其筛选结果为0，说明缺失`Description`变量值的**观测值**其`UnitPrice`变量值也为零。而`Description`表示产品名称，`UnitPrice`表示产品单价，如果二者同时缺失，我们认为其不具备商品交易实质。根据分析目标“挖掘畅销产品……”，我们认为这些观测值无法提供有效含义，因此对于这些观测值(这些条数据)后续可以进行删除。

对缺失数据的`CustomerID`变量重复上述操作后发现，有些`CustomerID`变量值缺失的观测值的其它变量值并不为零或空值。而`CustomerID`表示客户编号，不是分析“畅销产品”的必要变量，因为保留`CustomerID`变量值为缺失的观测值。

In [81]:
original_data[original_data["CustomerID"].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

### 4.2.2评估重复数据

根据数据变量的含义来看，虽然`InvoiceNo`,`StockCode`，`CustomerID	`都是唯一标识符，但是一次交易可能包含多件商品，因此`InvoiceNo`可以存在重复；不同交易可以包含同件商品，因此`StockCode`可以存在重复；顾客可以进行多次交易或下单多个商品，因此`CustomerID`也可以存在重复。

所以针对该数据集，我们无需评估重复数据。

### 4.2.3评估不一致数据

不一致数据可能存在于`country`变量中，我们要查看是否存在多个不同值指代同一国家的情况。

In [82]:
original_data["Country"].value_counts() #显示对于“数据集中Country变量的值”的种类数及其个数

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


从`Country`变量值来看，“USA”=“United States”=美国，“United Kingdom”=“UK”=“U.K.”=英国，因此在数据清洗步骤，需要对这些值进行统一，只保留一个指代值。

### 4.2.4评估无效或错误数据

可以通过DataFrame的`describe`方法，对数值统计信息进行快速了解

In [83]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

从输出结果来看，`Quantity`和`UnitPrice`存在负数，不符合实际，会对后续数值分析产生影响，因此，我们先筛选出变量`Quantity`和`UnitPrice`为负的观测值，进一步评估其含义。

In [84]:
original_data[original_data["Quantity"] < 0 ]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

In [85]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C") ]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

以上猜想被证实错误，因为还存在`Quantity`变量为负但`InvoiceNo`不以C开头的观察值。

但是根据以上输出结果，这些筛选出的观察值的`UnitPrice`观察值均为0，因此增加`UnitPrice`的条件进行验证。

In [86]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C") &(original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

根据输出结果，猜想得到验证，当`Quantity`变量为负时，观测值满足“`InvoiceNo`以CW为开头，表示订单被取消 ∀ `UnitPrice`为0，说明单位为0英镑”这些交易均不是有效成交数据，不贡献销售，不在后续分析范围内，因此我们将在数据清洗步骤，将`Quantity`为负数的观察值删除。

接下来，对`UnitPrice`进行同流程操作。

In [87]:
original_data[original_data["UnitPrice"] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

从输出结果来看，`UnitPrice`为负的观测值都是坏账调账交易，不是实际商业交易，因此也在数据清洗步骤将其删除。

# 5清洗数据

根据步骤4“评估数据”得到的结论，我们需要进行的数据清洗包括：

- 把`InvoiceDate`变量的数据类型转换为日期时间
- 把`CustomerID`变量的数据类型转换为字符串
- 把`Description`变量值缺失的观测值删除
- 把`Country`变量值“USA”替换为“United States”
- 把`Country`变量值“U.K.”、“UK”替换为“United Kingdom”
- 把`Quantity`变量值为负数的观察值删除
- 把`UnitPrice`变量值为负数的观察值删除

## 5.1创建原始数据集的副本操作

为了区分开“清洗后数据”和“原始数据”，我们创建新的变量`cleaned_data`,让它为`original_data`复制出的副本，我们之后的清洗步骤都将在`clean_data`数据集上操作。

In [88]:
clean_data = original_data.copy()  #复制“original_data”数据集为“clean_data”
clean_data.head()  #显示“clean_data”前几个观测值(前几行)            

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 5.2处理数据结构问题

## 5.3处理数据内容问题

把`InvoiceDate`变量的数据类型转换为日期时间类型

In [89]:
clean_data["InvoiceDate"] = pd.to_datetime(clean_data["InvoiceDate"])  # “对于clean_data数据集中的InvoiceDate变量(列)”重新赋值为“时间类型”
clean_data["InvoiceDate"]

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

把`CustomerID`变量的数据类型转换为字符串

In [90]:
clean_data["CustomerID"] = clean_data["CustomerID"].astype(str)
clean_data["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

把`CustmoerID`变量值结尾的“.0”删掉

In [91]:
clean_data["CustomerID"] = clean_data["CustomerID"].str.slice(0,-2)
clean_data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

把`Description`变量缺失的观测值(行)删除，并查看该变量剩余空缺值的个数。

In [92]:
clean_data.dropna(subset=['Description'],inplace=True)

In [93]:
clean_data["Description"].isnull().sum()

np.int64(0)

把`Country`变量值“USA”替换为“United States”,并检查替换后的“USA”个数

In [94]:
clean_data['Country'] = clean_data['Country'].replace({'USA': 'United States'})

In [95]:
len(clean_data[clean_data['Country'] == 'USA'])

0

把`Country`变量值“U.K.”、“UK”替换为“United Kingdom”,并检查替换后`UK`和`U.K.`变量值的个数：

In [96]:
clean_data['Country'] = clean_data['Country'].replace({'U.K.': 'United Kingdom','UK': 'United Kingdom'})
print(len(clean_data[clean_data['Country'] == 'UK']))
len(clean_data[clean_data['Country'] == 'U.K.'])

0


0

把`Quantity`变量值为负数的观察值删除,并检查替换后`Quantity`变量值为负数的个数

In [97]:
clean_data = clean_data[clean_data['Quantity'] >= 0]
len(clean_data[clean_data["Quantity"] < 0])

0

把`UnitPrice`变量值为负数的观察值删除,并检查替换后`UnitPrice`变量值为负数的个数

In [98]:
clean_data = clean_data[clean_data['UnitPrice'] >= 0]
len(clean_data[clean_data["UnitPrice"] < 0])

0

## 5.4保存已清洗后的整齐数据

完成数据清洗后，把干净整齐的数据保存到新的文件里，文件名为`e_commerce_cleaned.csv`,且不保留(0,1,2,3,……)的行索引。

In [99]:
clean_data.to_csv("e_commerce_cleaned.csv",index=False)

In [100]:
pd.read_csv("e_commerce_cleaned.csv").head()

C:\Users\11485\AppData\Local\Temp\ipykernel_8764\2686214642.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("e_commerce_cleaned.csv").head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice CustomerID         Country  
0  2010-12-01 08:26:00       2.55      17850  United Kingdom  
1  2010-12-01 08:26:00       3.39      17850  United Kingdom  
2  2010-12-01 08:26:00       2.75      17850  United Kingdom  
3  2010-12-01 08:26:00       3.39      17850  United Kingdom  
4  2010-12-01 08:26:00       3.39      17850  United Kingdom

# 6整理数据

# 7分析数据

# 8可视化分析结论